midterm-2023-11-06/code/5100-nenad-saldziev/script.ts

```diff
@@ -0,0 +1,188 @@
+
+document.addEventListener("DOMContentLoaded", backsitcode)
+
+let bands: metalBand[] = []; // all authors == bands
+
+async function backsitcode() {
+    const data = await waitData();
+    bands = data;
+
+    fillter(bands); // fillNationalities == fillter
+
+    displayBands(bands); // displayAuthors == displayBands
+
+    const nameSort = document.getElementById("sort-name")!;
+    nameSort.addEventListener("click", sortByName); // sortByName == sortByName
+
+    const idSort = document.getElementById("sort-id")!;
+    idSort.addEventListener("click", sortById); // sortById == sortById
+
+    const applyFilterButton = document.getElementById("apply-filter")!;
+    applyFilterButton.addEventListener("click", applyFilter); // applyFilter == applyFilter
+
+    const modal = document.getElementById("biblio-details")!;
+    modal.addEventListener("click", () => {
+        modal.classList.add("hidden");
+    });
+}
+
+const nameSorter: BandSorter = (first, second) => first.name.localeCompare(second.name);
+const idSorter: BandSorter = (first, second) => first.id - second.id;
+
+const sortByName = () => {
+    const sortedBands = bands.toSorted(nameSorter);
+    displayBands(sortedBands);
+}
+
+const sortById = () => {
+    const sortedBands = bands.toSorted(idSorter);
+    displayBands(sortedBands);
+}
+
+const fillter = (bands: metalBand[]) => {
+    const filter = document.getElementById("country-filter")!;
+
+    const countrys = new Set<string>();
+    for (const band of bands) {
+        countrys.add(band.location);
+    }
+
+    for (const country of countrys) {
+        const option = document.createElement("option");
+        option.value = country;
+        option.innerHTML = country;
+        filter.appendChild(option);
+    }
+}
+
+const applyFilter = () => {
+    const countryElement = document.getElementById("country-filter") as HTMLSelectElement;
+    const country = countryElement.value;
+
+    const aliveElement = document.getElementById("alive-filter") as HTMLSelectElement;
+    const alive = aliveElement.value;
+
+    let filteredBands = bands;
+    if (country !== "all") {
+        filteredBands = filteredBands.filter(band => band.location === country);
+    }
+    if (alive !== "all") {
+        filteredBands = filteredBands.filter(bands => {
+            if (alive === "yes") {
+                return bands.formed === undefined;
+            }
+            return !!bands.formed;
+        })
+    }
+
+    displayBands(filteredBands);
+
+
+
+    // if (nationality === "all") {
+    //     displayAuthors(authors);
+    //     return;
+    // }
+
+    // const filteredAuthors = authors.filter(author => author.nationality === nationality);
+    // displayAuthors(filteredAuthors);
+}
+
+const waitData = async () => {
+    const datajson = "https://raw.githubusercontent.com/sweko/uacs-internet-programming-exams/main/midterm-2023-11-06/data/bands.json"; // dataUri == datajson
```

Comment:
The loading of the data does not take into account the existence of the `metalBands` property in the data.